src/frontend/src/binder/expr/mod.rs

```diff
@@ -462,6 +469,68 @@ impl Binder {
         Ok(func_call.into())
     }
 
+    /// The helper function to check if the current case-when
+    /// expression in `bind_case` could be optimized
+    /// into `ConstantLookupExpression`
+    fn check_bind_case_optimization(
+        &mut self,
+        conditions: Vec<Expr>,
+        results_expr: Vec<ExprImpl>,
+        operand: Option<Box<Expr>>,
+        fallback: Option<ExprImpl>,
+        constant_lookup_inputs: &mut Vec<ExprImpl>,
+    ) -> bool {
+        if conditions.len() < CASE_WHEN_ARMS_OPTIMIZE_LIMIT {
+            return false;
+        }
+
+        // TODO(Zihao): we could possibly optimize some simple cases when
+        // `operand` is None in the future, the current choice is not conducting the optimization.
+        // e.g., select case when c1 = 1 then (...) when (same pattern) then (... ) [else (...)] end from t1;
+        if let Some(operand) = operand {
+            let Ok(operand) = self.bind_expr_inner(*operand) else {
+                return false;
+            };
+            constant_lookup_inputs.push(operand);
+        } else {
+            return false;
+        }
+
+        for (condition, result) in zip_eq_fast(conditions, results_expr) {
+            // If the result expression is not const
+            // we will also not conduct the optimization
+            if !result.is_const() {
+                return false;
+            }
```

Comment:
Why do we need the result to be a const? I think the lookup expression just needs the condition to be a const. As a result, we can lookup the corresponding arm with the operand to fetch the result expression.

Reply:
The initial reason for this is that, for the following case-when expression,
```sql
select case c1 when 1 then c1 + 1 else c1 - 1 from t1;
```
The input chunk will need to be split to single row in order to be evaluated.
But yes, I think it's okay to let this happen.